monai/engines/trainer.py

```diff
@@ -147,23 +151,32 @@ def _iteration(self, engine: Engine, batchdata: Dict[str, torch.Tensor]):
             kwargs: Dict = {}
         else:
             inputs, targets, args, kwargs = batch
+        # put iteration outputs into engine.state
+        engine.state.output = output = {Keys.IMAGE: inputs, Keys.LABEL: targets}
```

Comment:
the `_iteration` method is extended to support additional events (such as FORWARD_COMPLETED). line 155 updates the output dictionary `engine.state.output` before this method returns. is this a reasonable usage of ignite? @vfdev-5 could you please help review, many thanks!

Reply:
@wyli Yes, you are correct about pointing out that: `engine.state.output` will be overwritten when `_iteration` will return `output` again. Here, however as far as I understand the idea is to make `engine.state.output` accessible in handlers attached to your custom events: `IterationEvents.FORWARD_COMPLETED`, `IterationEvents.LOSS_COMPLETED` etc.
The same idea can be also realized by passing directly the data as : 
```python
def _iteration(...):

    output[Keys.PRED] = self.inferer(inputs, self.network, *args, **kwargs)
    engine.fire_event(IterationEvents.FORWARD_COMPLETED, output) 
    output[Keys.LOSS] = self.loss_function(output[Keys.PRED], targets).mean()
    engine.fire_event(IterationEvents.LOSS_COMPLETED, output)
...
def on_forward_completed(engine, output):
    batch = engine.state.batch  # your input batch
    assert Keys.PRED in output
    ...

def on_loss_completed(engine, y_pred, loss):
    batch = engine.state.batch  # your input batch
    assert Keys.PRED in output
    assert Keys.LOSS in output
    ...
```
In this case, you do not need to write `output` to the state knowing that it will be overwritten in the end...

Reply:
Hi @vfdev-5 ,

Thanks for your review and comments.
We added more events in the iteration, and want to **unify the data sharing** with handlers through `engine`.
For example, if user want to add some additional operations to `predictions` before loss computation and also add the noise to `engine.state.output`:
```py
class AddNoise:
    def attach(self, engine):
        engine.add_event_handler(Events.FORWARD_COMPLETED, self)

    def __call__(self, engine):
        noise = torch.rand(...)
        engine.state.output[Keys.PREDS] += noise
        engine.state.output[noise] = noise
```
And I think `engine.state.output` will not be overwritten at the end of this iteration as we return this `output` directly.
Your example code also can work for this case, but user needs to remember **different args for different event callbacks** in the handler, I feel that's not a unified API, so I want to always pass `engine` as handler arg, it's much easier for me to explain the API usage to users.
I am writing a detailed document to help users quickly understand how to use MONAI workflows which are based on ignite. Users who never used ignite complained many questions about the data-sharing part, especially not clear "what is exactly stored in the engine?" "what can I get in handler callback?", etc. We need to define a standard data-sharing method instead of ad-hoc solution for every single program.

Thanks.

Reply:
Thanks the explanation! Yeah, makes sense to have a unified API. Just a nit with this approach is that user also should remember what is already added to `state.output`.

> And I think engine.state.output will not be overwritten at the end of this iteration as we return this output directly.

Yes, by overwritten I wanted to say that the ref to `output` will be reassigned in the end to `output`... Anyway, yes, it is possible to do that.

> Users who never used ignite complained many questions about the data-sharing part, especially not clear "what is exactly stored in the engine?" "what can I get in handler callback?", etc. We need to define a standard data-sharing method instead of ad-hoc solution for every single program.

This is very interesting feedback! Could you please share more feedback on the API such that we could improve our docs 👍

Reply:
Hi @vfdev-5 ,

Thanks for your comments.
I am writing the document to make a standard data-sharing method based on `ignite.engine`, and will make a sheet to describe every element in `engine` and especially the items in `engine.state`, then users just need to check this sheet to read/write expected data in their handlers or add their own data to `engine.state`.
And we mainly recommend users to write their own handlers to do **customized logic** if the built-in `SupervisedTrainer` and existing handlers can't satisfy their reuqests.
The main purpose of this design to make the `data-sharing` through workflows much more straight-forward.
Sure, I will share the document in MONAI pages when it's ready.

Thanks.

Reply:
thanks for the discussion! still I feel setting `engine.state.output` twice (within `_iteration` and as the return value of `_iteration`) is a bit confusing and looks like a workaround solution. but I also could find a better way without making large changes. so I'll approve this PR for now.

Hi Nic, perhaps we should share the document in https://github.com/Project-MONAI/MONAI/tree/master/docs/source as it's important